scylla/src/transport/session_test.rs

```diff
@@ -166,6 +166,72 @@ async fn test_prepared_statement() {
     }
 }
 
+#[tokio::test]
+#[ignore]
+async fn test_batch() {
+    let session = Session::connect("127.0.0.1:9042", None).await.unwrap();
+
+    session.query("CREATE KEYSPACE IF NOT EXISTS ks WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 1}", &[]).await.unwrap();
+    session
+        .query("DROP TABLE IF EXISTS ks.t;", &[])
+        .await
+        .unwrap();
+    session
+        .query(
+            "CREATE TABLE IF NOT EXISTS ks.t (a int, b int, c text, primary key (a, b))",
+            &[],
+        )
+        .await
+        .unwrap();
+
+    // Wait for schema agreement
+    std::thread::sleep(std::time::Duration::from_millis(300));
+
+    let prepared_statement = session
+        .prepare("INSERT INTO ks.t (a, b, c) VALUES (?, ?, ?)")
+        .await
+        .unwrap();
+
+    use crate::batch::Batch;
+    let mut batch: Batch = Default::default();
+    batch.append_statement("INSERT INTO ks.t (a, b, c) VALUES (?, ?, ?)");
+    batch.append_statement("INSERT INTO ks.t (a, b, c) VALUES (7, 11, '')");
+    batch.append_statement(prepared_statement);
+
+    let values = [
+        values!(1_i32, 2_i32, "abc"),
+        Vec::new(),
+        values!(1_i32, 4_i32, "hello"),
+    ];
+
+    session.batch(&batch, &values).await.unwrap();
```

Comment:
How about changing the `append_statement` method so that it takes both statement/query string and parameters? It will be less error-prone - it will be harder to mismatch statements and query parameters, and we avoid the problem of query/values count mismatch. Besides, you can keep a `Vec` of `(BatchStatement, &[Value])`s and avoid the generic stuff in `request::batch::Batch`.

Reply:
Such a change would cause a need to recreate `Batch` each time we want to send the same statements, but with different values. In order not to lose flexibility, I suggest leaving `append_statement` method as it is, and adding new structure named e.g. `BatchBuilder`, that would  have `append_statement` method described by you, and `build` method returning `(Batch, Vec<&[Value]>)`.
```rust
let mut builder: BatchBuilder = Default::default();
builder.append_statement(
    "INSERT INTO ks.t (a, b, c) VALUES (?, ?, ?)",
    values!(1_i32, 2_i32, "abc")
);

let (batch, values) = builder.build();

session.batch(&batch, &values).await.unwrap();
```
What do you think about it?

Reply:
I looked at Cassandra drivers for other languages (Java, Python, Go), and they all seem to only expose an API which takes statements and values at the same time:

https://docs.datastax.com/en/developer/java-driver/4.0/manual/core/statements/batch/
https://docs.datastax.com/en/developer/python-driver/3.24/api/cassandra/query/#cassandra.query.BatchStatement
https://godoc.org/github.com/gocql/gocql#Batch.Query

I see your point about batch reusability, but I'm still not entirely convinced if separating statements from values is a good idea. I'll leave the decision about which APIs to keep to @psarna.

If we decide to do keep both, then we should use a different name than the `BatchBuilder` (maybe `BoundBatch`? I don't have any better suggestion for now, though).

Reply:
I think, that https://godoc.org/github.com/gocql/gocql#Batch.Bind is worth a look, as it provides a way of setting different values for associated statement each time batch is executed ([example](https://gist.github.com/havaker/3258fd6982eb0877296a625b7b23e0f5)).

Reply:
Hard to decide... both sides make sense:
 * it's good to have similar APIs to existing drivers
 * it's good to let the users reuse batch statements, replacing only the values

We definitely need both ways then. The API for the first option could also look like:
```rust
batch.append_bound_statement(statement, values!(v1, v2, v3);
```
... and then, when sending the batch, if we see a bound statement, we supply its bound values instead of trying to use the values provided by the user.

In any case, I'm also fine with leaving this pull request as is, as long as we create a new github issue for providing the option to bind values to each separate statement in a batch. @havaker please create such an issue and add a TODO in code which shortly describes that we also want to implement this second API in the future.

Reply:
Done, added TODO comment and created #108